src/sciline/_provider.py

```diff
@@ -170,10 +171,18 @@ def __init__(
         self, *, args: dict[str, Key], kwargs: dict[str, Key], return_: Optional[Key]
     ) -> None:
         """Build from components, use dedicated creation functions instead."""
+        # Duplicate type hints could be allowed in principle, but it makes structure
+        # analysis and checks more difficult and error prone. As there is likely
+        # little utility in supporting this, we disallow it.
+        if len(set(args.values()) | set(kwargs.values())) != len(args) + len(kwargs):
+            raise ValueError("Duplicate type hints found in args and/or kwargs")
         self._args = args
         self._kwargs = kwargs
         self._return = return_
 
+    def __len__(self) -> int:
```

Comment:
I don't think `__len__` is the best choice here. The user might wonder whether this includes the return annotation or not. (This is how `len(__annotations__)` works)

Reply:
By the same argument, isn't what existing methods like `keys()` return also confusing?

Reply:
True. It should be more like `argument_keys`.

Reply:
Since this is just an internal implementation detail, I suggest to leave this for now.